mesh/exodus_writer.cpp

```diff
@@ -0,0 +1,1075 @@
+   void GenerateExodusIIElementBlocks();
+
+   /// @brief Extracts boundary ids and determines the element IDs and side IDs
+   /// (Exodus II) for each boundary element.
+   void GenerateExodusIIBoundaryInfo();
+
+   /// @brief Iterates over the elements to extract a unique set of node IDs
+   /// (or vertex IDs if first-order).
+   std::unordered_set<int> GenerateUniqueNodeIDs();
+
+   /// @brief Populates vectors with x, y, z coordinates from mesh.
+   void ExtractVertexCoordinates(std::vector<double> & coordx,
+                                 std::vector<double> & coordy,
+                                 std::vector<double> & coordz);
+
+   /// @brief Writes node connectivity for a particular block.
+   /// @param block_id The block to write to the file.
+   void WriteNodeConnectivityForBlock(const int block_id);
+
+   /// @brief Writes boundary information to file.
+   void WriteBoundaries();
+
+   /// @brief Writes the block IDs to the file.
+   void WriteBlockIDs();
+
+   /// @brief Writes a title to the file.
+   void WriteTitle();
+
+   /// @brief Writes the number of elements in the mesh.
+   void WriteNumOfElements();
+
+   /// @brief Writes the floating-point word size (4 == float; 8 == double).
+   void WriteFloatingPointWordSize();
+
+   /// @brief Writes the API version.
+   void WriteAPIVersion();
+
+   /// @brief Writes the database version.
+   void WriteDatabaseVersion();
+
+   /// @brief Writes the maximum length of a line.
+   void WriteMaxLineLength();
+
+   /// @brief Writes the maximum length of a name.
+   void WriteMaxNameLength();
+
+   /// @brief Writes the number of blocks.
+   void WriteNumElementBlocks();
+
+   /// @brief Writes all element block parameters.
+   void WriteElementBlocks();
+
+   /// @brief Called by @a WriteElementBlockParameters in for-loop.
+   /// @param block_id Block to write parameters.
+   void WriteElementBlockParameters(int block_id);
+
+   /// @brief Writes the coordinates of nodes.
+   void WriteNodalCoordinates();
+
+   /// @brief Writes the file size (normal=0; large=1). Coordinates are specified
+   /// separately as components for large files (i.e. xxx, yyy, zzz) as opposed
+   /// to (xyz, xyz, xyz) for normal files.
+   void WriteFileSize();
+
+   /// @brief Writes the nodesets. Currently, we do not support nodesets.
+   void WriteNodeSets();
+
+   /// @brief Writes the mesh dimension.
+   void WriteMeshDimension();
+
+   /// @brief Writes the number of timesteps. Currently, we do not support
+   /// multiple timesteps.
+   void WriteTimesteps();
+
+   /// @brief Writes a dummy variable. This is to circumvent a bug in LibMesh where
+   /// it will skip the x-coordinate when reading in an ExodusII file if the id of
+   /// the x-coordinates is 0. To prevent this, we define a dummy variable before
+   /// defining the coordinates. This ensures that the coordinate variable IDs have
+   /// values greater than zero.
+   void WriteDummyVariable();
```

Comment:
This can't be fixed in LibMesh?

Reply:
It's an issue for libMesh users without HDF5 caused by a rather subtle bug in an older version of their exodus reader.

[#libMesh/3283](https://github.com/libMesh/libmesh/issues/3823)

Reply:
Thanks for the context. We can leave it in for now but should track if there's a fix in libmesh and remove it. I'd reference the github issue https://github.com/libMesh/libmesh/issues/3823 in the comment.